File: nim/src/sphere.nim
import std/options
import std/math

import vec3
import point3
import ray
import material

type
  Sphere* = object
    center*: Point3
    radius*: float
    material*: Material

func unitSphere*(): Sphere =
  Sphere(center: zerosVec3(), radius: 1.0)

func hit*(self: Sphere; ray: Ray; t_min, t_max: float): Option[HitRecord] =
  # -b +- sqrt(b ** 2 - 4*a*c)
  # --------------------------
  #          2*a
  let oc = ray.origin - self.center
  let a = dot(ray.direction, ray.direction)
  let halfB = dot(oc, ray.direction)
  let c = dot(oc, oc) - self.radius * self.radius

  let determinant = halfB * halfB - a * c
  if determinant < 0.0:
    return none(HitRecord)

  let sqrtd = sqrt(determinant)
  var root = (-halfB - sqrtd) / a
  if root < t_min or t_max < root:
    root = (-halfB + sqrtd) / a
    if root < t_min or t_max < root:
      return none(HitRecord)

  let p = ray.at(root)
  let normal = (p - self.center) / self.radius
  return some(qualifiedHitRecord(ray, p, normal, self.material, root))

File: rust/src/sphere.rs
use crate::hit_record::HitRecord;
use crate::material::Material;
use crate::point3::Point3;
use crate::ray::Ray;

pub struct Sphere {
    pub center: Point3,
    pub radius: f32,
    pub material: Material,
}

impl Sphere {
    pub fn hit(&self, ray: &Ray, t_min: f32, t_max: f32) -> Option<HitRecord> {
        let oc = ray.origin - self.center;
        let a = ray.direction.dot(ray.direction);
        let half_b = oc.dot(ray.direction);
        let c = oc.dot(oc) - self.radius * self.radius;

        let discriminant = half_b * half_b - a * c;
        if discriminant < 0.0 {
            return None;
        }

        let sqrtd = discriminant.sqrt();
        let mut root = (-half_b - sqrtd) / a;
        if root < t_min || t_max < root {
            root = (-half_b + sqrtd) / a;
            if root < t_min || t_max < root {
                return None;
            }
        }

        let p = ray.at(root);
        let normal = (p - self.center) / self.radius;
        Some(HitRecord::qualified_hit_record(
            ray,
            self.material,
            p,
            root,
            normal,
        ))
    }
}
